app/src/main/java/de/westnordost/streetcomplete/quests/crossing/AddCrossing.kt

```diff
@@ -0,0 +1,186 @@
+
+            val roads = roadsByNodeId.getValue(nodeId)
+            val neighbouringRoadPositions = roads
+                .flatMap { it.getNodeIdsNeighbouringNodeId(nodeId) }
+                .map { mapData.getNode(it)!!.position }
+
+            val neighbouringFootwayPositions = footways
+                .flatMap { it.getNodeIdsNeighbouringNodeId(nodeId) }
+                .map { mapData.getNode(it)!!.position }
+
+            /* So, surrounding the shared node X, in the simple case, we have
+             *
+             * 1. position A, B neighbouring the shared node position which are part of a road way
+             * 2. position P, Q neighbouring the shared node position which are part of the footway
+             *
+             * The footway crosses the road if P is on one side of the polyline spanned by A,X,B and
+             * Q is on the other side.
+             *
+             * How can a footway that has a shared node with a road not cross the latter?
+             * Imagine the road at https://www.openstreetmap.org/node/258003112 would continue to
+             * the south here, then, still none of those footways would actually cross the street
+             *  - only if it continued straight to the north, for example.
+             *
+             * The example brings us to the less simple case: What if several roads roads share
+             * a node at a crossing-candidate position, like it is the case at every T-intersection?
+             * Also, what if there are more than one footways involved as in the link above?
+             *
+             * We look for if there is ANY crossing, so all polylines involved are checked:
+             * For all roads a car can go through point X, it is checked if not all footways that
+             * go through X are on the same side of the road-polyline.
+             * */
+            val nodePos = mapData.getNode(nodeId)!!.position
+            val roadAngle1 = neighbouringRoadPositions[0].initialBearingTo(nodePos)
+            val anyFootwayCrossesAnyRoad = (1 until neighbouringRoadPositions.size).any { i ->
+                val roadAngle2 = nodePos.initialBearingTo(neighbouringRoadPositions[i])
+                val roadTakesRightTurn = (roadAngle2 - roadAngle1).normalizeDegrees(-180.0) > 0
+
+                val sides = neighbouringFootwayPositions.map {
+                    val angle = nodePos.initialBearingTo(it)
+                    val side1 = (angle - roadAngle1).normalizeDegrees(-180.0)
+                    val side2 = (angle - roadAngle2).normalizeDegrees(-180.0)
+                    if (roadTakesRightTurn) {
+                        if (side1 > 0 && side2 > 0) 1 else -1
+                    } else {
+                        if (side1 > 0 || side2 > 0) 1 else -1
+                    }
+                }
+                val anyAreOnDifferentSidesOfRoad = sides.toSet().size > 1
+                anyAreOnDifferentSidesOfRoad
+            }
+            anyFootwayCrossesAnyRoad
```

Comment:
what kind of syntax is being used here? I am really stumped how it works

Reply:
Quirks of Kotlin syntax I don't like neither. Imagine there is a `return@retainAll` there

Reply:
You could pull the lambdas out into a functions:

```kt
fun anyAreOnDifferentSidesOfRoad(i) {
    // ...
    return sides.toSet().size > 1
}

fun anyFootwayCrossesAnyRoad(nodeId, footways) {
    // ...
    return (1 until neighbouringRoadPositions.size).any(::anyAreOnDifferentSidesOfRoad)
}

// ...

footwaysByNodeId.entries.retainAll(::anyFootwayCrossesAnyRoad)
```

Reply:
I will look at that later, thanks

Reply:
Alright, I hope with this commit, the code is easier to read:

https://github.com/streetcomplete/StreetComplete/pull/2999/commits/a510cd525a23962202dd8e61f98fece7cc30ee44